### PR TITLE
Dedupe typeorigin

### PR DESCRIPTION
**Pull request by richo**
I've started on refactoring the error handling code to avoid the need to reparse generated errors in `span_*`, but would rather land this incrementally as one monolithic PR (and have un-fond memories of merge conflicts from various other monoliths)

r? @eddyb

### CODE REVIEW

File: src/librustc/middle/infer/mod.rs [hunk 1]
```diff
@@ -22,6 +22,7 @@ pub use middle::ty::IntVarValue;
 pub use self::freshen::TypeFreshener;
 pub use self::region_inference::GenericKind;
 
+use core::fmt;
```
Comment (eddyb) [inline src/librustc/middle/infer/mod.rs]: Isn't this exported through `std::fmt` as well?

File: src/librustc/middle/infer/mod.rs [hunk 2]
```diff
@@ -128,6 +129,30 @@ pub enum TypeOrigin {
     EquatePredicate(Span),
 }
 
+impl TypeOrigin {
+    fn as_str(&self) -> &str {
```
Comment (eddyb) [inline src/librustc/middle/infer/mod.rs]: This should return `&'static str`, because it can.

**Comment by eddyb** [comment]
LGTM, modulo nits.

**Comment by richo** [comment]
Nits addressed! Thanks for that.

r? @eddyb

**Comment by eddyb** [comment]
Thanks! @bors r+

**Pull request merged**